WW3_OUNF configurable metadata

ukmo-ccbunney opened the pull request:
Ref: Issue #32 

## Description
This PR provides provides a method to configure the netCDF metadata on a per-variable basis for `ww3_ounf`.
This is a requirement for the UKMO generate CMEMS compliant gridded output netCDF files.

## Method
Default meta data is set by the WW3 code (as per the original ww3_ounf module). These metadata entries can be overridden using the `ounfmeta.inp` text file with the following format:
```
$ comment line
$ (blank lines are ignored)
$ tabs are converted to spaces
$
  META [ IFI [ IFJ ]  |  FLDID ]   [ IFC ]
       attr_name = attr_value
       attr_name = attr_value
       extra_attr = extra_value [type]

  META .... repeat
  .
  .
```

Each META line selects a WW3 output field either by its `IFI`, `IFJ` (group, field) value, or the output `tag name` (e.g. HS, WND, etc), as defined in `ww3_shel`. 

An optional extra IFC value selects the component for multi-component fields (e.g. wind).

The attributes are then set using name = value pairs.
This can be one of the following:
   -  "varnm"
   -  "ename"
   -  "standard_name", or "varns"
   -  "long_name" or "varnl"
  -  "globwave_name" or "varng"
   -  "direction_reference", "dir_ref" or "varnd"
   -  "comment" or "varnc"
   -  "units"
   -  "valid_min" or "vmin"
   -  "valid_max" or "vmax"
   -  "scale_factor" or "fsc"

If and `attr_name` does not match one of those defined in the META_T, then it is assumed to be an option "user" or "extra" metadata pair. In this case, you can set an optional "type" to be "c", "i", or "f" for character, integer or real to specify how the value will be written to the netCDF file.

Meta data for partitioned parameters have a special <IPART> placeholder that will be merged in with the current partition number when output.

For "global attributes" the special `META global` entry can be used. In this case, up to 30 global attributes can be set using the same syntax as for the "extra" attributes.

Example input file:
```
 $ Lines starting with dollars are comments.
 $ The line starts a meta-data section for the depth field
 META DPT
   standard_name = depth
   long_name = "can be quoted string"
   comment = or an unquoted string
   vmax = 999.9
   $ Next attribute is a "user" or "extra" attribute (needs type)
   my_integer_attribute = 999 i

 $ Next one is HSig (group 2, field 1)
 META 2 1
   varns = "sig. wave height"
   varnl = "this is long name"

 $ Next one is second component of wind. It also sets an
 $ "extra" meta data value (height - a float)
 META WND 2
   standard_name = "v-wind"
   height = 10.0 "r"

$ Partitioned data can use the <IPART> marker to merge in the partition number during output
META PHS
   long_name = "partitioned sig. wv. ht for partition <IPART>"

$ Global metadata - uses "extra" field syntax - max 30 entrues
META global
  institution = UKMO
  comment "space separated strings should be quoted" c
  version = 1.0 r
```

## Notable code changes
  - All meta data previously set in that output loop in `ww3_ounf` is now stored in arrays of META_T derived types in the new `w3oundmetamd.ftn` module. These meta data is grouped in arrays of  "output group" and "output field", in a similar way to the groupings of output flags on `ww3_shel` and `ww3_ounf`. Each output variable can have up to 3 components (to handle cases such as wind which outputs two components; u-wind and v-wind). When each field is written, it's meta data is selected from these arrays using the IFI, IFJ and NFIELD loop variables.
  - NCVARTYPE has been moved to the new w3ounfmetamd.ftn module.
  - UNITS2 and FSC3 are no longer required as each component has its own specific meta data.
  - `valid_min` and `valid_max` are now set in the units of the field and scaled appropriately by the `scale_factor` when written out.
 - Partition parameters share meta data for each partition, but can use the `<IPART>` placeholder string to add in the current partition number. [A future update might also include other place holders for strings such as "primary", "secondary", etc.]
 - W3IOGOMD: The "field name" matching SELECT statements in `W3READFLGRD` and `W3FLGRDFLAG` have been factored out into a separate subroutine (`W3FLDTOIJ`). This is to facilitate the matching of field names to IFI,IFJ indices when reading in the meta data input file.
 - W3IOGOMD: I had to reinstate line 761 as it is actually used in `ww3_ounf`. However, there seems to be some confusion as to what the field actually is...
 - `W3ODATMD`: Small bug fix to correct the values stored in NOGE for groups 1 and 5.

## Difference in outputs
Output netCDF files are generally identical (when no user meta data is supplied), apart from the following cases which were bugs in the original code:
 - GlobWAVE names were being incorrectly written on the 3rd component of tensor fields (garbage string).
 - When variables are type REAL, the `vmin` and `vmax` were not correct (incorrect scaling)
 - Tensor fields did not have consistent scaling factors for SHORT variables on the 3rd component.

ukmo-ccbunney committed:
Initial changes ot split out meta data to seperate module

ukmo-ccbunney committed:
Added w3ncmetamd module to allow optional extra
meta data for each field.

Fixed issue where ENAME was not being set

Fixed issue where scaling factor FSC was only being
set on first pass.

ukmo-ccbunney committed:
Added w3nmlounfmeta.ftn - this module provides all netCDF meta data for
ww3_ounf via a set of namelists that enables runtime overriding of
values.

ukmo-ccbunney committed:
Commented out some temporary variables.
Added initialisation of AMETA (array of meta data)

ukmo-ccbunney committed:
renamed file

ukmo-ccbunney committed:
Added new version of w3nmlounfmeta.ftn (old move to _v1)
Required update to w3iogomd to factor out the "output flag" => "IFI,IFJ" functionality.

ukmo-ccbunney committed:
Bugfix to extra attribute writing.

ukmo-ccbunney committed:
- Changed the way meta data arrays are initialised.
  - Small change to w3odatmd to ensure NOGE(1) is correctly set when
	BT4, IS2 or SETUP switches are enabled.

ukmo-ccbunney committed:
Removed some old source files

ukmo-ccbunney committed:
Renamed ounf meta module

ukmo-ccbunney committed:
Moved NCVARTYPE from ww3_ounf to w3ounfmetamd module.

ukmo-ccbunney committed:
Cleaned up w3ounfmetamd a bit and added subroutine/module header
comments.
Also - had to reinstate IDOUT(5,11) in w3iogomd as that field is used in
ww3_outf (increased NOSE(5) to 11)

ukmo-ccbunney committed:
- Tidied up some stuff
  - removed some comments/debug.
  - Fixed typo in partitioned output metadata

ukmo-ccbunney committed:
Alternative input method for meta data using a structure text file
rather than a fortran namelist file

ukmo-ccbunney committed:
Added handling of TABS in input file.
Added subroutune header comments.

ukmo-ccbunney committed:
Removed some hidden control characters from the module comments (probably ended up there due a copy and paste)

ukmo-ccbunney committed:
- Added ability to specify "global" netCDF attributes.
  - Factored out some common functionality into seperate subroutines.
  - Fixed incorrect scale_value attrbute name (scale_factor)

ukmo-ccbunney committed:
Fixed some missing calls to CHECK_ERROR

ukmo-ccbunney committed:
Updated all error outputs.

ukmo-ccbunney committed:
Removed some unused variables and commented lines

ukmo-ccbunney committed:
VMIN and VMAX (valid_min and valid_max) are now read in via ounfmeta.inp
as floating point values representing the range in the units specified.
Prior to writing out, these values are scaled using "scale_factor" and
converted to integer values (as per the original ww3_ounf behaviour).

ukmo-ccbunney committed:
Finished refactoring w3iogomd (name to flag handling).

ukmo-ccbunney committed:
Updated revision numbers to 7.XX milestone

ukmo-ccbunney committed:
1. scale_factor and _FillValue now set in w3ounfmetamd.
  2. fixed bug where local copy of NCVARTYPE was being used
  3. meta data for all components of IFI,IFJ combination now loaded once and stored in an array. Avoids having to call GETMETA multiple times and also makes FSC3 redundant

ukmo-ccbunney committed:
Fixed incorrect min/max for WND

ukmo-ccbunney committed:
Fixed wrong min/max for WND (again!)

ukmo-ccbunney committed:
- Fixed bug where NCVARTYPE could be set incorrcetly due to it being a
   module variable. Use new NCVARTYPEI variable (original user value) in
   ww3_ounf.
 - Fixed some more incorrect metadata
 - Removed redundant variables from ww3_ounf.
 - Removed some commented out code

ukmo-ccbunney committed:
Tidying up.

ukmo-ccbunney committed:
Added manual entry

ukmo-ccbunney commented:
I will add a full matrix comparison output when it is complete.

ukmo-ccbunney committed:
Removed old unused file

ukmo-ccbunney commented:
Tagging @thesser1 and @CarstenHansen as interested parties.

ukmo-ccbunney committed:
Missing FLAGUNR USE import when RTD switch enabled.

ukmo-ccbunney committed:
Trim a string

ukmo-ccbunney committed:
Make ENAME 10 characters, as per original ww3_ounf, and modified
ww3_ounf screen output to keep consistent output for regtests.

ukmo-ccbunney commented:
# Regression test results

Full regression test results attached:
[matrixComp_PR266.zip](https://github.com/NOAA-EMC/WW3/files/5565807/matrixComp_PR266.zip)

## Summary of differences:
No change to any binary WW3 outputs.

### Pre-existing differences:
  - `mww3_test_03`: Known to not be bit-4-bit
  - `ww3_tp2.18/work_TIDE`: ww3_prtide_current.out - ordering of "BAD" output lines in file is arbitrary in MPI mode.

### Expected (new) differences
  - `ww3_ta1` and `ww3_ts1`: log.ww3 file showing output "Wind sea period" instead of "Undefined / Not used". See comment in description for PR regarding this.
  - Any test with netCDF output (vartype = real) have differences in `valid_min` and `valid_max` values - these are now correctly scaled to the data range.

ukmo-ccbunney committed:
Re-instated 20 char length for ENAME - extra lengxpht needed if <IPART>
placeholders are used. Set format to A10 in screen output to keep
log files comparible.

ukmo-ccbunney commented:
By the way - do we consider this addition requires a new regtest? I can add one if needed.

aliabdolali commented:
@ukmo-ccbunney It would be great if we add 1/2 additional regtests? Can we add it to 1/2 of the existing tests?

aliabdolali commented:
@ukmo-ccbunney It seems the tar file from ftp is not from the most upadated version. 
Althought you changed it in the previous staging, sth went wrong in develop and it still calls the old one
Please modify it in your PR in model/bin/ww3_from_ftp.sh:
_ww3ver=v7.07
wget https://ftp.emc.ncep.noaa.gov/static_files/public/WW3/ww3_from_ftp.${ww3ver}.tar.gz_

ukmo-ccbunney commented:
> 
> 
> @ukmo-ccbunney It would be great if we add 1/2 additional regtests? Can we add it to 1/2 of the existing tests?

OK - I'll add a regtest for this. I think one should suffice. I can't see a good place to put it in any of the existing tests, so I'll just create a new one (`ww3_tnc1`?)

ukmo-ccbunney committed:
Updating file version and name of FTP tar file.

ukmo-ccbunney commented:
> Please modify it in your PR in model/bin/ww3_from_ftp.sh:
> ww3ver=v7.07
> wget [https://ftp.emc.ncep.noaa.gov/static_files/public/WW3/ww3_from_ftp.${ww3ver}.tar.gz](https://ftp.emc.ncep.noaa.gov/static_files/public/WW3/ww3_from_ftp.$%7Bww3ver%7D.tar.gz)

Done. :)

aliabdolali commented:
> > @ukmo-ccbunney It would be great if we add 1/2 additional regtests? Can we add it to 1/2 of the existing tests?
> 
> OK - I'll add a regtest for this. I think one should suffice. I can't see a good place to put it in any of the existing tests, so I'll just create a new one (`ww3_tnc1`?)

Great, I'll update the wiki about the new test.

ukmo-ccbunney committed:
Added regtest for configurable netCDF metadata in ww3_ounf

ukmo-ccbunney commented:
New regression test added (ww3_tnc1).
Required small modification to `run_test` script to ensure the `ounfmeta.inp` is correctly linked in.

mickaelaccensi commented on model/ftn/w3ounfmetamd.ftn:
```diff
@@ -0,0 +1,2893 @@
+      ! 3. Set the default values for the OUNF netCDF meta data.
+      CALL DEFAULT_META()
+
+      ! If the ounfmeta.nml exists, read this in to override defaults:
```
should it be ounfmeta.inp ?

ukmo-ccbunney replied:
Yes - I've changed it.

mickaelaccensi commented on model/ftn/w3ounfmetamd.ftn:
```diff
@@ -0,0 +1,2893 @@
+!
+!  2. Method:
+!
+!     The internal WW3 field can be speicifed either as an [IFI, IFJ]
```
typo : specified

ukmo-ccbunney replied:
Corrected.

mickaelaccensi commented on model/ftn/w3ounfmetamd.ftn:
```diff
@@ -0,0 +1,2893 @@
+!  2. Method:
+!
+!     The internal WW3 field can be speicifed either as an [IFI, IFJ]
+!     integer combination, or a field ID tag (such as "HS").
```
is it case sensitive ?

ukmo-ccbunney replied:
At the moment, the "META" tag and the "field names" are case sensitive. I did wonder about making them case-insensitive; do you think it would be better? I shouldn't be too difficult to do (assuming an ASCII character set).

mickaelaccensi replied:
in w3servmd there is a function STR_TO_UPPER that could call to automatically convert it to upper case.

ukmo-ccbunney replied:
Perfect! I will use that.

ukmo-ccbunney replied:
OK - I've made the META keyword and the "field names IDs" case insensitive in the ounfmeta.inp file: 2612878b

mickaelaccensi commented on model/ftn/w3ounfmetamd.ftn:
```diff
@@ -0,0 +1,2893 @@
+      REAL :: RVAL
+!/
+      ERR = NF90_PUT_ATT(NCID, VARID, 'long_name', META%VARNL)
+      IF(ERR /= NF90_NOERR) RETURN
```
what exactly happen if there is an error when putting attribute in netcdf file ? it does not stop with an error message ?

ukmo-ccbunney replied:
The WRITE_META subroutine returns an error code if a netCDF error occurs - this is then captured by the call to CHECK_ERR on lines 2465 and 2601 in ww3_ounf.ftn.

mickaelaccensi replied:
ok

mickaelaccensi commented on model/ftn/w3ounfmetamd.ftn:
```diff
@@ -0,0 +1,2893 @@
+      IF(ERR /= NF90_NOERR) RETURN
+      
+      ! Note: Valid min/max are scaled by scale_factor 
+      !       and converted to integers.
```
so even for a NF90_FLOAT variable, the valid min and max are integers ?

ukmo-ccbunney replied:
That's how it was before, but you're right - I should only be scaling to integer for NF90_SHORT data types.
I will update so if NCVARTYPE is 2, it is scaled and converted to integer, otherwise it is written straight out as a float.

mickaelaccensi replied:
great

ukmo-ccbunney replied:
OK - done for commit 45f56ef4

mickaelaccensi commented on model/ftn/w3ounfmetamd.ftn:
```diff
@@ -0,0 +1,2893 @@
+          CASE('i')
+            READ(GLOBAL_META(I)%ATTVAL, *) IVAL
+            ERR = NF90_PUT_ATT(NCID, VARID, GLOBAL_META(I)%ATTNAME, IVAL)
+            IF(ERR /= NF90_NOERR) RETURN
```
same question, should it stops with an error message instead of "RETURN" ?

ukmo-ccbunney replied:
Same answer as above, but in this case it is missing the CHECK_ERR after the subroutine call - oops! I will add one.

mickaelaccensi commented on model/ftn/w3ounfmetamd.ftn:
```diff
@@ -0,0 +1,2893 @@
+      TYPE(META_T), POINTER :: META(:)
+!
+!----------GROUP 1 ----------------
+!
```
could you add a comment to specify that VMIN and VMAX values should be set as is. Scale factor will be applied in another subroutine
so by default, here it should be a REAL value for VMIN and VMAX, isn't ?

ukmo-ccbunney replied:
Comment added to subroutine header.
Yes - VMIN and VMAX are reals, although if the user enters an integer, it will still be read as a real.

ukmo-ansaulter replied:
In which case I think we could be brave and remove the commented out scaled VMIN/VMAX? They've been useful for reviewing the change but will be redundant once this is approved

ukmo-ccbunney replied:
OK - done.

mickaelaccensi commented on model/ftn/ww3_ounf.ftn:
```diff
@@ -35,6 +36,8 @@
 !/    18-Jun-2020 : Support for 360-day calendar.       ( version 7.08 )
 !/    07-Oct-2019 : RTD option with standard lat-lon
 !/                  grid when nesting to rotated grid   ( version 7.11 )
+!/    03-Nov-2020 : Moved NetCDF metadata to seperate   ( version 7.XX )
```
typo :separate ? but since you wrote it in many comments I feel it's probably not a typo ?

ukmo-ccbunney replied:
No, you are right - it is a typo! It's just that I consistently spell it incorrectly!
Good spot (it's always a bit embarrassing when a non-native English speaker corrects a native English speakers spelling!)

mickaelaccensi replied:
I'm always disappointed between US and UK spelling so that's why I've checked on googletranslate to be sure !

mickaelaccensi commented on model/ftn/ww3_ounf.ftn:
```diff
@@ -981,33 +982,17 @@
             FLDIR = .FALSE.
-            UNITS2='SAME'
             IF (NCVARTYPEI.EQ.3) NCVARTYPE=2
 !
```
it would be good to keep in comment the default name of the variable even if it can be changed by the user in the metadata

ukmo-ccbunney replied:
I don't quite understand this comment - do you want me to add the UNITS2 line back in, but leave it commented out?

mickaelaccensi replied:
my request was to add the variable standard name before each test "ELSE IF ( IFI .EQ. 2 .AND. IFJ .EQ. 4 ) THEN"   to remember on which variable the test is applied

ukmo-ccbunney replied:
Ah - I see. Yes, I will do that.

ukmo-ccbunney replied:
Done in commit 7f6e462a

mickaelaccensi commented on model/ftn/ww3_ounf.ftn:
```diff
@@ -5104,6 +3596,9 @@
       ENDIF
 
+      ! ChrisB: Write user global attributes:
+      CALL WRITE_GLOBAL_META(NCID, IRET)
```
should you add a CHECK_ERROR call here ?

ukmo-ccbunney replied:
Yup! Done!

mickaelaccensi reviewed (commented):
really awesome work and clearly written ! We definitely need to apply this method for ww3_ounp and ww3_trnc !

ukmo-ccbunney committed:
Changes from review:
   - Fixed some typos.
   - Updated some comments.
   - Added missing CHECK_ERR call.

ukmo-ccbunney committed:
valid_min and valid_max are now floats when output variable type is 4
(float)

ukmo-ccbunney committed:
Made META keyword and field name strings case insensitive in
ounfmeta.inp file

mickaelaccensi commented:
it's all good for me, I approve the PR.
Once Andy gave his comments, then it would be good to redo the matrix to be sure there is no bug inserted with the last commits

ukmo-ccbunney committed:
Added output name as comment on each IF-ELSEIF entry for clarity.

ukmo-ccbunney commented:
> it's all good for me, I approve the PR.
> Once Andy gave his comments, then it would be good to redo the matrix to be sure there is no bug inserted with the last commits

Many thanks Mickael.
I'll make sure to rerun the regtests.

aliabdolali commented:
@ukmo-ccbunney please let me know once you feel happy about it, then I'll check regtests on our HPC. 
AA

ukmo-ansaulter commented on manual/run/ww3_ounf.tex:
```diff
@@ -7,14 +7,112 @@ \subsubsection{Gridded output NetCDF post-processor} \label{sec:ww3ounf}
+no type is supplied, it defaults to a characer type. Valid
+types are one of ["c", "r", "i"] for character/string, 
+real/float or integer values respectively.
+
```
Might be worth adding a note that setting a string attribute to blank ("") will prevent the attribute being written out; e.g. if we wanted to switch the globwave name off

ukmo-ccbunney replied:
Agreed. Done.

ukmo-ccbunney replied:
Actually - just to clarify, this is only true for the "globwave", "comment" and "direction_reference" attributes; all other "default" attributes are always written out (as they are required for CF compliance). I have noted this in the manual.

ukmo-ansaulter commented on model/ftn/w3odatmd.ftn:
```diff
@@ -758,7 +760,8 @@
       IDOUT( 5,10)  = 'Dominant break prob '
-!      IDOUT( 5,11)  = 'Breaker passage rate'
+      IDOUT( 5,11)  = 'Wind sea period' ! C.Bunney - reinstated this as is used in ww3_ounf
+                                        ! Is it suposed to be defunct? It is not in ww3_outf...
```
This parameter is not currently listed in the manual? However, I think @ukmo-juan-castillo has mentioned it previously as a coupled model requirement? Suggest this isn't addressed here, but worth resolving in issue #206

ukmo-juan-castillo replied:
> This parameter is not currently listed in the manual? However, I think @ukmo-juan-castillo has mentioned it previously as a coupled model requirement? Suggest this isn't addressed here, but worth resolving in issue #206

This parameter is not required for coupling at the moment, and I do not think it will.

mickaelaccensi replied:
wind sea period was introduced by Fabrice Ardhuin. If it's not explained in the manual, it would be good to create an issue to remember to do it

ukmo-ccbunney replied:
Also - it is an output field in ww3_ounf, but NOT in ww3_outf.

Confusingly, though, in field name matching subroutine in w3iogomd, the internal field name for IFI=5, IFJ=11 is `FWS`, even though it writes out field `TWS` (line 917)

`TWS` is actually IFI=4, IFJ=16 (line 872) - bit this seems wrong as later this is associated with partitioned wind sea fraction (line 2987).

I think those two parameters have become switched at some point.

I agree that an issue needs raising to sort out this! I don't want to change that stuff here as it will affect my regtests and it feels like a separate bug fix.

EDIT: I will raise an Issue for this.

ukmo-ccbunney replied:
Issue raised for the TWS/FWS issue here: https://github.com/NOAA-EMC/WW3/issues/267

ukmo-ansaulter commented on model/ftn/w3ounfmetamd.ftn:
```diff
@@ -0,0 +1,2923 @@
+      META(1)%UNITS  = '1'
+      META(1)%ENAME  = '.wbt'
+      META(1)%VARNM='wbt'
+      META(1)%VARNL='domiant wave breaking probability'
```
"domiant" should be "dominant"?

ukmo-ccbunney replied:
Yes - fixed (not one of my typos this time!)

ukmo-ansaulter reviewed (approved):
Nice work @ukmo-ccbunney and thanks @mickaelaccensi for the thorough first review

Have added a couple of minor comments - but all looks good to me

ukmo-ansaulter reviewed (approved):
Added another approving review, as my last attempt didn't seem to remove the merging block - but should have done?

ukmo-ansaulter commented:
> Added another approving review, as my last attempt didn't seem to remove the merging block - but should have done?

Ah - it's because I'm a moron and hadn't noticed this was directly in the NOAA repo, not the UKMO one.... (add slappy forehead emoji as appropriate)

ukmo-ccbunney committed:
Manual entry updated

ukmo-ccbunney committed:
Typo correction

ukmo-ccbunney committed:
Removed some commented out vmin/vmax entries

ukmo-ccbunney committed:
Fixed format statememt to make ww3_ounf log output consistent with
previous version.

ukmo-ccbunney commented:
@aliabdolali - rerun of regtest matrix after changes from review is consistent (slightly better in fact as I fixed a whitespace difference in the ww3_ounf.out logs). Results are attached here: 

[matrixComp_PR266_v2.zip](https://github.com/NOAA-EMC/WW3/files/5590534/matrixComp_PR266_v2.zip)

Expected differences are as detailed in the PR description.

I think this PR is ready for you to run your regtests now!

aliabdolali commented:
Hi @ukmo-ccbunney 
I have the following error during **ww3_ounf** execution for **ww3_tnc1** test.

_./bin/run_test -b slurm -c hera -S -T -o netcdf ../model ww3_tnc1_

   _Processing /scratch2/COASTAL/coastal/save/Ali.Abdolali/Source/Nov28UKMET/regtests/ww3_tnc1/input/ww3_ounf.inp
   Screen output routed to /scratch2/COASTAL/coastal/save/Ali.Abdolali/Source/Nov28UKMET/regtests/ww3_tnc1/work/ww3_ounf.out
forrtl: severe (59): list-directed I/O syntax error, unit -5, file Internal List-Directed Read
Image              PC                Routine            Line        Source             
ww3_ounf           000000000063B7BE  for__io_return        Unknown  Unknown
ww3_ounf           000000000066588D  for_read_int_lis_     Unknown  Unknown
ww3_ounf           0000000000663FF6  for_read_int_lis      Unknown  Unknown
ww3_ounf           00000000006167F8  Unknown               Unknown  Unknown
ww3_ounf           00000000006268AF  Unknown               Unknown  Unknown
ww3_ounf           000000000040C1DA  Unknown               Unknown  Unknown
ww3_ounf           000000000040AD1E  Unknown               Unknown  Unknown
libc-2.17.so       00007F1CE8A86555  __libc_start_main     Unknown  Unknown
ww3_ounf           000000000040AC29  Unknown               Unknown  Unknown

ERROR: Error occured during /scratch2/COASTAL/coastal/save/Ali.Abdolali/Source/Nov28UKMET/model/exe/ww3_ounf execution_

ukmo-ccbunney commented:
> 
> 
> Hi @ukmo-ccbunney
> I have the following error during **ww3_ounf** execution for **ww3_tnc1** test.
> 
> _./bin/run_test -b slurm -c hera -S -T -o netcdf ../model ww3_tnc1_
> 
> _Processing /scratch2/COASTAL/coastal/save/Ali.Abdolali/Source/Nov28UKMET/regtests/ww3_tnc1/input/ww3_ounf.inp Screen output routed to /scratch2/COASTAL/coastal/save/Ali.Abdolali/Source/Nov28UKMET/regtests/ww3_tnc1/work/ww3_ounf.out forrtl: severe (59): list-directed I/O syntax error, unit -5, file Internal List-Directed Read Image PC Routine Line Source ww3_ounf 000000000063B7BE for__io_return Unknown Unknown ww3_ounf 000000000066588D for_read_int_lis_ Unknown Unknown
> ww3_ounf 0000000000663FF6 for_read_int_lis Unknown Unknown
> ww3_ounf 00000000006167F8 Unknown Unknown Unknown
> ww3_ounf 00000000006268AF Unknown Unknown Unknown
> ww3_ounf 000000000040C1DA Unknown Unknown Unknown
> ww3_ounf 000000000040AD1E Unknown Unknown Unknown
> libc-2.17.so 00007F1CE8A86555 __libc_start_main Unknown Unknown
> ww3_ounf 000000000040AC29 Unknown Unknown Unknown
> 
> ERROR: Error occured during /scratch2/COASTAL/coastal/save/Ali.Abdolali/Source/Nov28UKMET/model/exe/ww3_ounf execution_

Hi Ali,
Any chance you could rerun that test with debugging flags turned on?
It looks like a problem with an internal read from a string, but I am not sure where it is happening.
Is it the Intel compiler you are using?
Chris.

aliabdolali commented:
Hi @ukmo-ccbunney 
I did it with intel_debug again (I had to remove _fpe0_ from _model/bin/cmplr.env_ to avoid _forrtl: error (73): floating divide by zero_): 

./bin/run_test -b slurm -c intel_debug -S -T -o netcdf ../model ww3_tnc1

   Processing /scratch2/COASTAL/coastal/save/Ali.Abdolali/Source/Nov28UKMET/regtests/ww3_tnc1/input/ww3_ounf.inp
   Screen output routed to /scratch2/COASTAL/coastal/save/Ali.Abdolali/Source/Nov28UKMET/regtests/ww3_tnc1/work/ww3_ounf.out
forrtl: severe (59): list-directed I/O syntax error, unit -5, file Internal List-Directed Read
Image              PC                Routine            Line        Source             
ww3_ounf           0000000000E830BE  Unknown               Unknown  Unknown
ww3_ounf           0000000000EAD18D  Unknown               Unknown  Unknown
ww3_ounf           0000000000EAB8F6  Unknown               Unknown  Unknown
ww3_ounf           0000000000D6F595  w3ounfmetamd_mp_r         783  w3ounfmetamd.F90
ww3_ounf           0000000000D6C1EE  w3ounfmetamd_mp_r         530  w3ounfmetamd.F90
ww3_ounf           0000000000D65072  w3ounfmetamd_mp_i         239  w3ounfmetamd.F90
ww3_ounf           00000000004122F3  MAIN__                    407  ww3_ounf.F90
ww3_ounf           000000000040AF9E  Unknown               Unknown  Unknown
libc-2.17.so       00007FD7FDAE5555  __libc_start_main     Unknown  Unknown
ww3_ounf           000000000040AEA9  Unknown               Unknown  Unknown


see w3ounfmetamd.F90 attached
[w3ounfmetamd.F90.zip](https://github.com/NOAA-EMC/WW3/files/5616336/w3ounfmetamd.F90.zip)

ukmo-ccbunney committed:
- Fixed typo in ounfmeta.inp (wrong closing quote).
 -Updated w3ounfmetamd to:
   - Handle empty attribute values more robustly (intel compiler needs an empty string to be specifed as "")
   - Better error reporting if internal list directed read fails.

ukmo-ccbunney committed:
Added blank globawave_entry to test empty attribute value processing

ukmo-ccbunney commented:
@aliabdolali - This looks like it is due to a mismatching closing quote in my ounfmeta.inp file:
```
standard_name="eastward_wind'
```
I've fixed this.

I also did some testing with the Intel compiler and I noticed that IFORT does not like to read from an empty string (Cray Fortran and GNU Fortran seem ok). I.e. it doesn't like this "empty" attribute line:
```
   varnm=
```
but it is happy with this:
```
   varnm=""
```

I have made a small update to `w3ounfmetamd` handle this as well as better error reporting of internal list directed reads.

The regtests are passing for me on GNU, Intel and Cray fortran compilers now.

aliabdolali commented:
Hi @ukmo-ccbunney 
Thanks again for doing this great work in a short period of time. I ran the regtests on our HPC with intel compiler successfully. The differences are the same as yours, as expected. 
Here is the comparison with the base.

[matrixcompNOAA.zip](https://github.com/NOAA-EMC/WW3/files/5626487/matrixcompNOAA.zip)

@JessicaMeixner-NOAA could you review this PR?

JessicaMeixner-NOAA commented:
@aliabdolali will review later today or tomorrow.

JessicaMeixner-NOAA reviewed (approved):
Looks good to me!

ukmo-ccbunney commented:
Thanks @JessicaMeixner-NOAA.
@aliabdolali - the Milestone for this change is still set at 7.XX. Should we get a version number associated with this now?

aliabdolali commented:
@ukmo-ccbunney Good point, the version number is 7.12

ukmo-ccbunney commented:
@aliabdolali  - how shall we deal with this? Shall I update this branch?

aliabdolali commented:
@ukmo-ccbunney could you change the version number, following https://github.com/NOAA-EMC/WW3/wiki/Code-Management and send a PR?

ukmo-ccbunney commented:
> 
> 
> @ukmo-ccbunney could you change the version number, following https://github.com/NOAA-EMC/WW3/wiki/Code-Management and send a PR?

Done: https://github.com/NOAA-EMC/WW3/pull/271
